src/gridcoin/voting/result.cpp

```diff
@@ -864,18 +864,38 @@ class VoteCounter
     {
         CTransaction tx;
 
-        if (!m_txdb.ReadDiskTx(txid, tx)) {
-            LogPrint(LogFlags::VOTE, "%s: failed to read vote tx", __func__);
+        bool read_tx_success = false;
+
+        // This is very ugly. In testing for implement poll expiration reminders PR2716, there is an issue with ReadDiskTx
+        // on very fast machines, where upon receipt of a vote on an existing poll, the poll builder tests for the transaction
+        // BEFORE it is committed to disk. This retry loop is essentially zero overhead for an immediate success, but does
+        // up to 10 tries over up to 2.5 seconds total to "wait" for the transaction to appear in leveldb.
+        for (unsigned int i = 0; i < 10; ++i) {
+            if (m_txdb.ReadDiskTx(txid, tx)) {
+                read_tx_success = true;
+                break;
+            } else {
+                LogPrintf("WARN: %s: failed to read vote tx in try %u", __func__, i + 1);
+            }
+
+            if (!MilliSleep(250)) {
+                // Interrupt with throw if MilliSleep interrupted by op sys signal.
+                throw InvalidVoteError();
+            }
+        }
+
+        if (!read_tx_success) {
+            LogPrintf("WARN: %s: failed to read vote tx after 10 tries", __func__);
```

Comment:
@div72 this is really horrible. I tried to track down the source of this and originally thought it was at line 1622 in main.cpp (ConnectBlock)...

'
            // Validate any contracts published in the transaction:
            if (!tx.GetContracts().empty()) {
                if (!CheckContracts(tx, mapInputs, pindex->nHeight)) {
                    return false;
                }

                int DoS = 0;
                if (block.nVersion >= 11 && !GRC::BlockValidateContracts(pindex, tx, DoS)) {
                    return tx.DoS(DoS, error("%s: invalid contract in tx %s, assigning DoS misbehavior of %i",
                                             __func__,
                                             tx.GetHash().ToString(),
                                             DoS));
                }
            }

            if (!ConnectInputs(tx, txdb, mapInputs, mapQueuedChanges, posThisTx, pindex, true, false))
                return false;
        }

        mapQueuedChanges[hashTx] = CTxIndex(posThisTx, tx.vout.size());
    }

    if (IsResearchAgeEnabled(pindex->nHeight)
        && !GridcoinConnectBlock(block, pindex, txdb, stake_value_in, nStakeReward, nFees))
    {
        return false;
    }

    pindex->nMoneySupply = ReturnCurrentMoneySupply(pindex) + nValueOut - nValueIn;

    if (!txdb.WriteBlockIndex(CDiskBlockIndex(pindex)))
        return error("%s: WriteBlockIndex for pindex failed", __func__);
'

Note here that under cs_main, the Gridcoin contract validation occurs, which also fires the contract handlers (which also signal a new poll or vote), and then the queued transactions are committed to disk shortly later.

In the earlier PR to improve polling performance, the the poll refresh does NOT block on cs_main, so I thought this spacing would possible create the condition where the queued transactions are not actually in leveldb by the time the signal fires and the roll result builder attempts to access the vote transaction.

To try and eliminate this, I tried putting a global scope critical section lock to protect this interval and required the LOCK in FetchVoteCandidate, instead of the horrible multiple try loop above, but it did not resolve the issue, so it must be elsewhere. I am worried this is actually deep within leveldb. Supposedly leveldb is supposed to be threading coherant on batch commits, but this would indicate that is not the case.

Reply:
Ok. See https://github.com/gridcoin-community/Gridcoin-Research/pull/2716/commits/1926b0668a80429d9228974a1d0390519c25720e and the attached log excerpt.

Note that threadnames in the logfile have been turned on here.
[poll_refresh_excerpt_debug.log](https://github.com/gridcoin-community/Gridcoin-Research/files/13531658/poll_refresh_excerpt_debug.log)

We clearly have a problem with leveldb.
...
2023-12-01T20:10:48Z [grc-msghand] INFO: ConnectBlock: cs_tx_val_commit_to_disk locked
2023-12-01T20:10:48Z [grc-msghand] INFO: CheckResearchReward: Post CheckReward: m_total_claimed = 42.36266666, research_owed = 32.26666666, out_stake_owed = 10.00, mrc_staker_fees = 0.00, mrc_fees = 0.00, mrc_rewards = 0.00
2023-12-01T20:10:48Z [grc-msghand] INFO: DetectReorg: registry_traversal_in_progress = 0, reorg_occurred_during_reg_traversal = 0,
2023-12-01T20:10:48Z [grc-msghand] INFO: DetectReorg: Setting reorg_occurred_during_reg_traversal to false.
2023-12-01T20:10:48Z [grc-msghand] INFO: AddVote: Added vote 088c14fc082118ef6a8b6d7b53feb04580f32eb61c827a311e62df38170be8dc to poll pr2716 test, poll_ref->Votes().size() = 17.
2023-12-01T20:10:48Z [grc-msghand] INFO: ConnectBlock: cs_tx_val_commit_to_disk unlocked
2023-12-01T20:10:48Z [grc-msghand] NotifyTransactionChanged e2b217b37d436c96a69e0c6b3034f505767d2327b7bb5204151a93f6355c165a status=2
2023-12-01T20:10:48Z [grc-msghand] NotifyTransactionChanged e2b217b37d436c96a69e0c6b3034f505767d2327b7bb5204151a93f6355c165a status=0
2023-12-01T20:10:48Z [gridcoinresearch-main] INFO: refresh: m_refresh_mutex trylock succeeded.
2023-12-01T20:10:48Z [PollTableModel_refresh] INFO: GetTimes: timer buildPollTable: Begin buildPollTable: elapsed time: 0 ms, time since last check: 0 ms.
2023-12-01T20:10:48Z [PollTableModel_refresh] INFO: GetTimes: timer buildPollTable: Begin BuildPollItem: elapsed time: 0 ms, time since last check: 0 ms.
2023-12-01T20:10:48Z [PollTableModel_refresh] INFO: GetTimes: timer buildPollTable: Begin BuildFor: elapsed time: 0 ms, time since last check: 0 ms.
2023-12-01T20:10:48Z [gridcoinresearch-main] updateWallet e2b217b37d436c96a69e0c6b3034f505767d2327b7bb5204151a93f6355c165a 2
2023-12-01T20:10:48Z [PollTableModel_refresh] INFO: BuildFor: number of votes = 17 for poll PR2716 test
2023-12-01T20:10:48Z [PollTableModel_refresh] INFO: GetTimes: timer buildPollTable: Begin CountVotes: elapsed time: 1 ms, time since last check: 1 ms.
2023-12-01T20:10:48Z [PollTableModel_refresh] INFO: FetchVoteCandidate: cs_tx_val_commit_to_disk locked
2023-12-01T20:10:48Z [PollTableModel_refresh] WARN: FetchVoteCandidate: failed to read vote tx.
2023-12-01T20:10:48Z [PollTableModel_refresh] INFO: FetchVoteCandidate: cs_tx_val_commit_to_disk unlocked
2023-12-01T20:10:48Z [PollTableModel_refresh] FetchVoteCandidate: tx has no vote contract
2023-12-01T20:10:48Z [PollTableModel_refresh] INFO: CountVotes: skipped invalid vote: 088c14fc082118ef6a8b6d7b53feb04580f32eb61c827a311e62df38170be8dc
2023-12-01T20:10:48Z [PollTableModel_refresh] INFO: FetchVoteCandidate: cs_tx_val_commit_to_disk locked
2023-12-01T20:10:48Z [PollTableModel_refresh] INFO: FetchVoteCandidate: cs_tx_val_commit_to_disk unlocked

cs_tx_val_commit_to_disk unlocked is logged in the grc_msghand thread, which indicates that UpdateTxIndex(hash, index)) has been called and is finished.

The lock of cs_tx_val_commit_to_disk originating from the signal occurs after this, but the tx read still fails. This indicates that UpdateTxIndex(hash, index) is returning BEFORE the information is committed to disk (or alternatively in the leveldb buffer associated with the name-value pairs so that any other reading thread would receive the results post the UpdateTxIndex call. This is an isolation failure in ACID.

I will now go look at how we have the environment set up in leveldb and the leveldb documentation to see if we have something set incorrectly in the leveldb environment.

Reply:
Ok. There is something more subtle going on here. Because other nodes running the same code that are not the source node for the vote update properly. So there has to be something about the transaction already being present in the mempool (on the source node for the vote) and how the queued transaction changes are dealt with.

Reply:
Ok. https://github.com/gridcoin-community/Gridcoin-Research/pull/2716/commits/c49b2f2bca056638800cb076564b3fbf961bf49b fixes the problem. I had the lock scoped incorrectly. It has to be held on the main side for the duration of the leveldb batch write ("transaction") and so must be taken and released in ReorganizeChain.

[poll_refresh_excerpt_debug_fixed.log](https://github.com/gridcoin-community/Gridcoin-Research/files/13537862/poll_refresh_excerpt_debug_fixed.log)

Reply:
@div72 I think this fix is good enough. We can do more refactoring later, but this change is simpler and doesn't mess with consensus critical code. To start moving GridcoinConnectBlock around, etc. requires a lot more thought.